pkg/alicloud/client/types.go

```diff
@@ -61,10 +61,22 @@ type ECS interface {
 	CheckIfImageExists(ctx context.Context, imageID string) (bool, error)
 	CheckIfImageOwnedByAliCloud(imageID string) (bool, error)
 	ShareImageToAccount(ctx context.Context, regionID, imageID, accountID string) error
-	DescribeSecurityGroups(request *ecs.DescribeSecurityGroupsRequest) (response *ecs.DescribeSecurityGroupsResponse, err error)
-	DescribeSecurityGroupAttribute(request *ecs.DescribeSecurityGroupAttributeRequest) (response *ecs.DescribeSecurityGroupAttributeResponse, err error)
-	DescribeKeyPairs(request *ecs.DescribeKeyPairsRequest) (response *ecs.DescribeKeyPairsResponse, err error)
+	DescribeSecurityGroups(request *ecs.DescribeSecurityGroupsRequest) (*ecs.DescribeSecurityGroupsResponse, error)
+	DescribeSecurityGroupAttribute(request *ecs.DescribeSecurityGroupAttributeRequest) (*ecs.DescribeSecurityGroupAttributeResponse, error)
+	DescribeKeyPairs(request *ecs.DescribeKeyPairsRequest) (*ecs.DescribeKeyPairsResponse, error)
 	DetachECSInstancesFromSSHKeyPair(keyName string) error
+	DescribeInstances(request *ecs.DescribeInstancesRequest) (*ecs.DescribeInstancesResponse, error)
+	DescribeAvailableResource(request *ecs.DescribeAvailableResourceRequest) (*ecs.DescribeAvailableResourceResponse, error)
+	RunInstances(request *ecs.RunInstancesRequest) (*ecs.RunInstancesResponse, error)
+	DeleteInstance(request *ecs.DeleteInstanceRequest) (*ecs.DeleteInstanceResponse, error)
+	CreateSecurityGroup(request *ecs.CreateSecurityGroupRequest) (*ecs.CreateSecurityGroupResponse, error)
+	DeleteSecurityGroup(request *ecs.DeleteSecurityGroupRequest) (*ecs.DeleteSecurityGroupResponse, error)
+	AllocatePublicIpAddress(request *ecs.AllocatePublicIpAddressRequest) (*ecs.AllocatePublicIpAddressResponse, error)
+
+	AuthorizeSecurityGroup(request *ecs.AuthorizeSecurityGroupRequest) (*ecs.AuthorizeSecurityGroupResponse, error)
+	RevokeSecurityGroup(request *ecs.RevokeSecurityGroupRequest) (*ecs.RevokeSecurityGroupResponse, error)
+	RevokeSecurityGroupEgress(request *ecs.RevokeSecurityGroupEgressRequest) (*ecs.RevokeSecurityGroupEgressResponse, error)
+	AuthorizeSecurityGroupEgress(request *ecs.AuthorizeSecurityGroupEgressRequest) (*ecs.AuthorizeSecurityGroupEgressResponse, error)
```

Comment:
These interfaces are declared here but the actual implementations lie elsewhere in the actuator of bastion controller. Note that the ECS interface here is not just a caller for the actual ecs client, it also wraps some of the common logic within. I would suggest that you move the implementations from the actuator here.

Reply:
The interface implementations from `vendor/xxx/aliyun` directly, take a example `RunInstances`  same as `DetachECSInstancesFromSSHKeyPair` which I was refer to.   the ECS interface is a caller from an actual ecs client if I understand correctly. so no need to move it is it?
<img width="1163" alt="image" src="https://user-images.githubusercontent.com/42234376/161510013-bed4a9f8-1f5f-4792-8f7d-94d8051c138d.png">.

Reply:
The alicloud client as it is now takes parameters like imageID, regionID, etc., and assembles respective actual AliCloud requests. So caller (user of our own client, for example, controller implementations) does not need to handle the raw requests by themselves. We do this to keep AliCloud client code in one place (/pkg/alicloud/client).

Reply:
committed it in e38bf733